server/src/main/java/org/opensearch/index/compositeindex/datacube/startree/utils/CoordinatedDocumentReader.java

```diff
@@ -0,0 +1,90 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.index.compositeindex.datacube.startree.utils;
+
+import org.apache.lucene.search.DocIdSetIterator;
+import org.opensearch.common.annotation.ExperimentalApi;
+
+/**
+ * Coordinates the reading of documents across multiple DocIdSetIterators.
+ * It encapsulates a single DocIdSetIterator and maintains the latest document ID and its associated value.
+ * @opensearch.experimental
+ */
+@ExperimentalApi
+public class CoordinatedDocumentReader {
```

Comment:
SequentialDocValuesIterator ? 

In the context of this class, there is no coordination happening , it just tracks docId & docValue - why not make the nextValue return the next value based on the doc id ? 

Lets move that logic from adapter to this reader.